atc/db/worker_cache.go

```diff
@@ -0,0 +1,278 @@
+			cache.removeWorker(workerName)
+		} else {
+			logger.Info("unexpected-operation", lager.Data{"operation": event.Operation})
+		}
+	}
+}
+
+func (cache *WorkerCache) listenContainers(notifications <-chan Notification) {
+	logger := cache.logger.Session("listen-containers")
+
+	for notification := range notifications {
+		if !notification.Healthy {
+			logger.Info("notification-unhealthy-will-refresh")
+			cache.ensureRefresh()
+			continue
+		}
+
+		var event TriggerEvent
+		if err := json.Unmarshal([]byte(notification.Payload), &event); err != nil {
+			logger.Error("invalid-payload", err)
+			continue
+		}
+
+		workerName, ok := event.Data["worker_name"]
+		if !ok {
+			logger.Info("missing-name")
+			continue
+		}
+
+		if event.Operation == "INSERT" {
+			cache.addWorkerContainerCount(workerName, 1)
+		} else if event.Operation == "DELETE" {
+			cache.addWorkerContainerCount(workerName, -1)
+		} else {
+			logger.Info("unexpected-operation", lager.Data{"operation": event.Operation})
+		}
+	}
+}
+
+func (cache *WorkerCache) refreshWorkerData() error {
+	cache.mut.Lock()
+	defer cache.mut.Unlock()
+
+	if !cache.needsRefresh() {
+		return nil
+	}
+
+	cache.logger.Debug("refreshing")
+
+	workers, err := getWorkers(cache.conn, workersQuery)
+	if err != nil {
+		return err
+	}
+
+	cache.workers = make(map[string]Worker, len(workers))
+	for _, worker := range workers {
+		cache.workers[worker.Name()] = worker
+	}
+
+	rows, err := psql.Select("worker_name, COUNT(*)").
+		From("containers").
+		Where("build_id IS NOT NULL").
+		GroupBy("worker_name").
+		RunWith(cache.conn).
+		Query()
+	if err != nil {
+		return err
+	}
+
+	defer Close(rows)
+
+	countByWorker := make(map[string]int, len(workers))
+
+	for rows.Next() {
+		var workerName string
+		var containersCount int
+
+		err = rows.Scan(&workerName, &containersCount)
+		if err != nil {
+			return err
+		}
+
+		countByWorker[workerName] = containersCount
+	}
+
+	cache.lastRefresh = time.Now()
+	cache.workerContainerCounts = countByWorker
+
+	return nil
+}
+
+func (cache *WorkerCache) needsRefresh() bool {
+	return time.Since(cache.lastRefresh) >= cache.refreshInterval
+}
+
+func (cache *WorkerCache) ensureRefresh() {
+	cache.mut.Lock()
+	defer cache.mut.Unlock()
+
+	cache.lastRefresh = time.Time{}
+}
+
+func (cache *WorkerCache) removeWorker(name string) {
+	cache.mut.Lock()
+	defer cache.mut.Unlock()
+
+	delete(cache.workers, name)
+	delete(cache.workerContainerCounts, name)
+}
+
+func (cache *WorkerCache) upsertWorker(name string) error {
+	worker, found, err := getWorker(cache.conn, workersQuery.Where(sq.Eq{"w.name": name}))
+	if err != nil {
+		return err
+	}
+	if !found {
+		// worker disappeared while trying to fetch, so ensure it's gone
+		cache.removeWorker(name)
+		return nil
+	}
+
+	cache.mut.Lock()
+	defer cache.mut.Unlock()
+
+	cache.workers[name] = worker
+	if _, ok := cache.workerContainerCounts[name]; !ok {
+		cache.workerContainerCounts[name] = 0
```

Comment:
when cache is un-sync with DB like this, should it be removed like line 259 or set to 0?

Reply:
This condition isn't necessarily about being out-of-sync - if a new worker is inserted, it won't yet have an entry in the `workerContainerCounts` table. Since it can't have containers yet, we add it to the table with 0 containers